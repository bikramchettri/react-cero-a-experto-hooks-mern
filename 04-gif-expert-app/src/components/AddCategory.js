import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({setCategories}) => {
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
        console.log('handleInputChange trabajando');
    }
    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length > 2){
            setCategories( categorias => [inputValue, ...categorias]); 
            setInputValue('');
        }
    }
    return (
        <form onSubmit={handleSubmit}>
        <p>{inputValue}</p>
                <input
                type="text"
                value = {inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )
}

AddCategory.propType = {
    setCategories: PropTypes.func.isRequired
}

