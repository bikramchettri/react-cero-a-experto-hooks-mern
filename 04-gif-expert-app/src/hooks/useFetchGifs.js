import { useState, useEffect } from 'react';
import {getGifs} from '../helpers/GetGifs'

export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => { 
        //se ejecuta una sola vez
        getGifs(category)
            .then( imgs => {
                setState({
                    data: imgs,
                    loading: false
                })
            });
    },[category]);

    return state; // { data:[], loading: true };
}