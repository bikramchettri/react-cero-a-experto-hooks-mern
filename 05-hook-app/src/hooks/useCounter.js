import { useState } from 'react';

export const useCounter = (initialState = 10) => {
    const [counter, setCounter] = useState(initialState); //state es el valor actual, setStave cuando se vaya a cambiar

    const increment = () => {
        setCounter( counter + 1);
    }

    const decrement = () => {
        setCounter( counter - 1);
    }

    const reset = () => {
        setCounter(initialState);
    }

    return {
        counter,
        increment,
        decrement,
        reset
    };
    
}
