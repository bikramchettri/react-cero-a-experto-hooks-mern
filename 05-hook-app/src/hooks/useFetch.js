import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {

    const isMouted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        return () => {
            isMouted.current = false;
        }
    }, []);

    useEffect(() => {
        setState({data: null, loading: true, error: null});
        fetch(url)
            .then( resp => resp.json())
            .then( data => {
                if(isMouted.current){ // si mouted es true cargo el setState
                    setState({
                        loading: false,
                        error: null,
                        data
                    });
                }    
            });
    }, [url]);

    return state;
}
