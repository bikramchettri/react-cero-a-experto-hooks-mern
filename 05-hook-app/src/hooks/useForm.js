import { useState } from 'react';

export const useForm = (initialState = {}) => {
    const [values, setValues] = useState(initialState); //values es el valor actual, setStave cuando se vaya a cambiar

    const handleInputChange = ({target}) => {
        
        setValues({
            ...values,
            [target.name]:target.value
        });
        
    }
 
    return [values, handleInputChange]; // retorno un arreglo
    
}
