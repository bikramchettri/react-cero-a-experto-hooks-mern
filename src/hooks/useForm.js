import { useState } from 'react';

export const useForm = (initialState = {}) => {
    const [values, setValues] = useState(initialState); //values es el valor actual, setStave cuando se vaya a cambiar

    const reset = () => { 
        setValues(initialState);
    };
    
    const handleInputChange = ({target}) => {
        
        setValues({
            ...values,
            [target.name]:target.value
        });
        
    }
    return [values, handleInputChange, reset]; // retorno un arreglo
    
}