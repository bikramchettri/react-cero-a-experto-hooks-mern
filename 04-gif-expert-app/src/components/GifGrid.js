import React from 'react'
import  GifGridItem  from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({category}) => {
    //const [images, setImages] = useState([]); //images el arreglo, setImages la funcion que lo llena
    const { data:images, loading } = useFetchGifs(category);
    
    return (
        <>
        <h3 className="animate__animated animate__flash">{category}</h3>
        {loading && <p>Cargando</p>}
             <div className="card-grid">
                
                {
                    images.map( img => ( 
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                    ))
                }
            </div> 
        </>
    )
}
