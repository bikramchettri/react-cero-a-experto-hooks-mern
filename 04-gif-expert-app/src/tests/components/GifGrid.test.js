import React from 'react';
import { shallow } from 'enzyme';
import {GifGrid} from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el componente <GifGrid/>', ()=> {

    test('debe de mostrar  <GifGrid/> correctamente ', () => {
        
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const category='One piece';
        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
     });

     test('debe de mostrar items cuando se cargan imagenes con useFetchGifs', () => {
         const gifs = [
             {
                 id: 'ABC',
                 url:'https://sdgfg.com',
                 title:'cualquier titulo'
             },
             {
                id: 'ABC1',
                url:'https://sdgkdkdfg.com',
                title:'cualquier titulo 2'
            }
         ]
         useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        const category='One piece';
        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);

     });
     
     
})