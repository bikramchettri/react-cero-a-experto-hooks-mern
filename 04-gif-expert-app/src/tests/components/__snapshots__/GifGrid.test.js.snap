// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrid/> debe de mostrar  <GifGrid/> correctamente  1`] = `
<Fragment>
  <h3
    className="animate__animated animate__flash"
  >
    One piece
  </h3>
  <p>
    Cargando
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas en el componente <GifGrid/> debe de mostrar items cuando se cargan imagenes con useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__flash"
  >
    One piece
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="cualquier titulo"
      url="https://sdgfg.com"
    />
    <GifGridItem
      id="ABC1"
      key="ABC1"
      title="cualquier titulo 2"
      url="https://sdgkdkdfg.com"
    />
  </div>
</Fragment>
`;
